blockchain/rolling_merkle.go

```diff
@@ -0,0 +1,208 @@
+package blockchain
+
+import (
+	"math"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+)
+
+// RollingMerkleTree computes the double SHA256 merkle root over a set of leaf
+// hashes. Interior nodes are opportunistically computed as new leaves are
+// added to the tree, consolidating any nodes that have both a left and right
+// child before permitting another leaf to be added. As a result, a
+// RollingMerkleTree only requires O(log n) additional storage to compute the
+// final root.
+type RollingMerkleTree struct {
+	// nodes contains all nodes in the tree. The nodes may correspond to
+	// interior nodes if their children have already been pruned.
+	nodes map[uint32]chainhash.Hash
```

Comment:
Wouldn't this be better served as a slice? So we just index in rather than having the overhead of a map. We'd also gain memory locality for all the items as well.